File: src/main/java/fr/ynerant/leveleditor/api/editor/RawMap.java
package fr.ynerant.leveleditor.api.editor;

import fr.ynerant.leveleditor.editor.Map;

import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class RawMap {
    private List<RawCase> cases;
    private java.util.Map<Integer, RawCase> cases_map;
    private int width;
    private int height;
    private transient BufferedImage font;

    public static RawMap create(List<RawCase> cases, int width, int height) {
        RawMap rm = new RawMap();
        rm.cases = cases;
        rm.width = width;
        rm.height = height;
        return rm;
    }

    public static RawMap create(Map map) {
        RawMap raw = new RawMap();
        raw.width = map.getWidth();
        raw.height = map.getHeight();
        raw.cases = new ArrayList<>();
        for (Case c : map.getAllCases()) {
            RawCase rc = RawCase.create(c);
            raw.cases.add(rc);
        }
        return raw;
    }

    public List<RawCase> getCases() {
        return cases;
    }

    public RawCase getCase(int x, int y) {
        if (cases_map == null) {
            cases_map = new HashMap<>();
            for (RawCase c : getCases())
                cases_map.put(c.getPosY() * width + c.getPosX(), c);
        }

        return cases_map.get(y * getWidth() + x);
    }

    public Collection<RawCase> getNeighbours(RawCase c) {
        List<RawCase> list = new ArrayList<>();
        list.add(getCase(c.getPosX() - 1, c.getPosY()));
        list.add(getCase(c.getPosX(), c.getPosY() - 1));
        list.add(getCase(c.getPosX() + 1, c.getPosY()));
        list.add(getCase(c.getPosX(), c.getPosY() + 1));
        return list.stream().filter(_c -> _c != null && _c.getCollision() == Collision.ANY).collect(Collectors.toList());
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public BufferedImage getFont() {
        return font;
    }

    public void setFont(BufferedImage font) {
        this.font = font;
    }
}

File: src/main/scala/fr/ynerant/leveleditor/api/editor/RawMap.scala
package fr.ynerant.leveleditor.api.editor

import java.awt.image.BufferedImage

import fr.ynerant.leveleditor.editor.GMap
import net.liftweb.json._


object RawMap {
	def create(cases: List[RawCase], width: Int, height: Int): RawMap = {
		new RawMap(cases, width, height)
	}

	def create(map: GMap): RawMap = {
		val raw = new RawMap(Nil, map.getWidth, map.getHeight)
		map.getAllCases.foreach(c => raw.cases ::= RawCase.create(c))
		raw
	}
}

case class RawMap(var cases: List[RawCase], var width: Int, var height: Int) {
	private var cases_map = null: Map[Integer, RawCase]
	private var font = null: BufferedImage

	def getNeighbours(c: RawCase): Iterable[RawCase] = {
		var list = Nil: List[RawCase]
		list ::= getCase(c.getPosX, c.getPosY - 1)
		list ::= getCase(c.getPosX + 1, c.getPosY)
		list ::= getCase(c.getPosX, c.getPosY + 1)
		list ::= getCase(c.getPosX - 1, c.getPosY)
		list ::= getCase(c.getPosX - 1, c.getPosY + 1)
		list ::= getCase(c.getPosX - 1, c.getPosY - 1)
		list ::= getCase(c.getPosX + 1, c.getPosY + 1)
		list ::= getCase(c.getPosX + 1, c.getPosY - 1)
		list.filter((_c: RawCase) => _c != null)
	}

	def getCase(x: Int, y: Int): RawCase = {
		if (cases_map == null) {
			cases_map = Map()
			getCases.foreach(c => cases_map = cases_map.updated(c.getPosY * width + c.getPosX, c))
		}
		cases_map.getOrElse(y * getWidth + x, null)
	}

	def getCases: List[RawCase] = cases

	def getWidth: Int = width

	def getHeight: Int = height

	def getFont: BufferedImage = font

	def setFont(font: BufferedImage): Unit = {
		this.font = font
	}

	override def toString: String = {
		implicit val formats: DefaultFormats.type = DefaultFormats
		Serialization.writePretty(this)
	}
}
